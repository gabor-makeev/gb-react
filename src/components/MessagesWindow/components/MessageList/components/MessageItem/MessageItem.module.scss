@import "src/styles/variables/colors";

.message {
  align-self: flex-end;
  margin-left: 36%;
  position: relative;

  & > .message__content {
    display: block;
    padding: 15px 50px 18px;
    border-radius: 25px;
    font-size: 26px;
    line-height: 30px;
    letter-spacing: 0.05em;
  }

  &__auth-user {
    padding-right: 21px;
    &:after {
      content: '';
      position: absolute;
      right: 0;
      bottom: 0;
      width: 18px;
      height: 18px;
      border-radius: 9px;
      background-color: $darkSeaGreen;
    }
    & > .message__content {
      background-color: $darkSeaGreen;
      color: $policeBlue;
      margin-bottom: 7px;
    }
  }

  &__other-user {
    align-self: flex-start;
    margin-left: 0;
    margin-right: 36%;
    padding-left: 14px;
    &:before {
      content: '';
      position: absolute;
      left: 0;
      bottom: 0;
      width: 18px;
      height: 18px;
      border-radius: 9px;
      background-color: $steelTeal;
    }
    & > .message__content {
      background-color: $steelTeal;
      color: $darkSeaGreen;
      margin-bottom: 10px;
    }
  }

  &__system-background {
    & > .message__content {
      background-color: $regularRed;
      color: $policeBlue;
    }
  }
}