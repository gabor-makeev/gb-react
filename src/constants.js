export const AUTHORS = {
  user: 'Gabor',
  bot: 'ðŸ¤– Bot',
};

export const STYLES = {
  color: {
    primary: '#FF4F4F',
    secondary: '#31E865',
  },
};
