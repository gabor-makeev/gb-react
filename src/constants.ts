import { Authors, Messages } from './default-types';

export const api = 'https://api.spaceflightnewsapi.net/v3/articles';

export const NAVIGATION = [
  {
    id: 1,
    name: 'Home',
    path: '/',
  },
  {
    id: 2,
    name: 'Profile',
    path: '/profile',
  },
  {
    id: 3,
    name: 'Messenger',
    path: '/messenger',
  },
];

export const DUMMY_CONTENT = {
  messages: [
    {
      body: 'placerat duis ultricies lacus sed turpis tincidunt id aliquet risus',
      userEmail: Authors.USER,
      userName: Authors.USER,
    },
    {
      body: 'et molestie ac feugiat sed lectus vestibulum mattis ullamcorper velit',
      userEmail: Authors.USER,
      userName: Authors.USER,
    },
    {
      body: 'felis eget velit aliquet sagittis id consectetur purus ut faucibus',
      userEmail: Authors.USER,
      userName: Authors.USER,
    },
  ] as Messages,
};
