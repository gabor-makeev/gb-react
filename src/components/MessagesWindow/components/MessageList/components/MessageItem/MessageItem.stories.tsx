import { MessageItem } from './MessageItem';
import { DUMMY_CONTENT } from 'src/constants';
import { ComponentMeta, ComponentStory } from '@storybook/react';

export default {
  title: 'MessageItem',
  component: MessageItem,
} as ComponentMeta<typeof MessageItem>;

const dummyMessage = DUMMY_CONTENT.messages[0];

const Template: ComponentStory<typeof MessageItem> = (args) => (
  <MessageItem {...args} message={dummyMessage} />
);

export const Primary = Template.bind({});
