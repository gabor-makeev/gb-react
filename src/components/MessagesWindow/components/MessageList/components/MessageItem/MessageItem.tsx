import { FC } from 'react';
import classNames from 'classnames';
import style from './MessageItem.module.scss';
import { Message } from 'src/default-types';

interface MessageItemProps {
  userEmail: string;
  message: Message;
}

export const MessageItem: FC<MessageItemProps> = ({ userEmail, message }) => {
  const MessageListItemClasses = classNames(style.message, {
    [style['message__system-background']]: !message.userEmail,
    [style['message__auth-user']]:
      message.userEmail && message.userEmail === userEmail,
    [style['message__other-user']]:
      message.userEmail && message.userEmail !== userEmail,
  });

  return (
    <li className={MessageListItemClasses} data-testid={'messageItem'}>
      <span className={style['message__content']}>{message.body}</span>
    </li>
  );
};
