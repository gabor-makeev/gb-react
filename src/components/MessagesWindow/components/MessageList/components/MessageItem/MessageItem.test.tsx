import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { MessageItem } from './MessageItem';
import { DUMMY_CONTENT } from '../../../../../../constants';

const dummyMessage = DUMMY_CONTENT.messages[0];

describe('Message', () => {
  it('should render', () => {
    render(<MessageItem message={dummyMessage} />);
  });

  it('should render text in <li>-tag', () => {
    render(<MessageItem message={dummyMessage} />);
    expect(screen.getByText(dummyMessage.text)).toBeInTheDocument();
  });

  it('should render text in <span>-tag', () => {
    render(<MessageItem message={dummyMessage} />);
    expect(screen.getByText(dummyMessage.author)).toBeInTheDocument();
  });

  // the code displayed below has to be deleted as soon as there is no way to test non-inline styles with jest:

  // it('should use default messageStyle property when such not passed', () => {
  //   render(<MessageItem message={dummyMessage} />);
  //   expect(screen.getByText(dummyMessage.text)).toHaveStyle(
  //     `background-color: ${STYLES.color.secondary}`
  //   );
  // });

  // the code displayed below has to be deleted as soon as there is no way to test non-inline styles with jest:

  // it('should use padding style passed in padding prop', () => {
  //   render(<MessageItem message={dummyMessage} padding={'100px 100px'} />);
  //
  //   expect(screen.getByText(dummyMessage.text)).toHaveStyle(
  //     'padding: 100px 100px'
  //   );
  // });

  it('should render with snapshot', () => {
    const { asFragment } = render(<MessageItem message={dummyMessage} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
