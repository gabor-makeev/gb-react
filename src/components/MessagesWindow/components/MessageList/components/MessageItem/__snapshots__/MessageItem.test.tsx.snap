// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should render with snapshot 1`] = `
<DocumentFragment>
  <li
    class="message message__other-user"
    data-testid="messageItem"
  >
    <span
      class="message__content"
    >
      placerat duis ultricies lacus sed turpis tincidunt id aliquet risus
    </span>
  </li>
</DocumentFragment>
`;
